subsys/dfu/boot/mcuboot.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2016-2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_DOMAIN "fota/mcuboot"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_FOTA_LEVEL
+#include <logging/sys_log.h>
+
+#include <stddef.h>
+#include <errno.h>
+#include <string.h>
+#include <flash.h>
+#include <zephyr.h>
+#include <init.h>
+
+#include "boot/mcuboot.h"
+
+/*
+ * Helpers for image trailer, as defined by mcuboot.
+ * Image trailer consists of sequence of fields:
+ *   u8_t copy_done
+ *   u8_t pading_1[BOOT_MAX_ALIGN - 1]
+ *   u8_t image_ok
+ *   u8_t pading_2[BOOT_MAX_ALIGN - 1]
+ *   u8_t magic[16]
+ */
+
+/* Strict defines: Defines in block below must be equal to coresponding
+ * mcuboot defines */
+#define BOOT_MAX_ALIGN 8
+#define BOOT_MAGIC_SZ  16
+#define BOOT_FLAG_SET 0x01
+#define BOOT_FLAG_UNSET 0xff
+/* end_of Strict defines */
+
+#define BOOT_MAGIC_GOOD  1
+#define BOOT_MAGIC_BAD   2
+#define BOOT_MAGIC_UNSET 3
+
+#define BOOT_FLAG_IMAGE_OK 0
+#define BOOT_FLAG_COPY_DONE 1
+
+#define COPY_DONE_OFFS(bank_offs) (bank_offs + FLASH_BANK_SIZE -\
+				   BOOT_MAGIC_SZ - BOOT_MAX_ALIGN * 2)
+
+#define IMAGE_OK_OFFS(bank_offs) (bank_offs + FLASH_BANK_SIZE - BOOT_MAGIC_SZ -\
+				  BOOT_MAX_ALIGN)
+#define MAGIC_OFFS(bank_offs) (bank_offs + FLASH_BANK_SIZE - BOOT_MAGIC_SZ)
+
+const u32_t boot_img_magic[4] = {
+	0xf395c277,
+	0x7fefd260,
+	0x0f505235,
+	0x8079b62c,
+};
+
+static struct device *flash_dev;
+
+static int boot_flag_offs(int flag, u32_t bank_offs, u32_t *offs)
+{
+	switch (flag) {
+	case BOOT_FLAG_COPY_DONE:
+		*offs = COPY_DONE_OFFS(bank_offs);
+		return 0;
+	case BOOT_FLAG_IMAGE_OK:
+		*offs = IMAGE_OK_OFFS(bank_offs);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int boot_write_flag(int flag, u32_t bank_offs)
+{
+	u8_t buf[FLASH_MIN_WRITE_SIZE];
+	u32_t offs;
+	int rc;
+
+	rc = boot_flag_offs(flag, bank_offs, &offs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	memset(buf, 0xFF, sizeof(buf));
+	buf[0] = BOOT_FLAG_SET;
+
+	flash_write_protection_set(flash_dev, false);
+	rc = flash_write(flash_dev, offs, buf, sizeof(buf));
+	flash_write_protection_set(flash_dev, true);
+
+	return rc;
+}
+
+static int boot_flag_get(int flag, u32_t bank_offs)
+{
+	u32_t offs;
+	int rc;
+	u8_t flag_val;
+
+	rc = boot_flag_offs(flag, bank_offs, &offs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	rc = flash_read(flash_dev, offs, &flag_val, sizeof(flag_val));
+	if (rc != 0) {
+		return rc;
+	}
+
+	return flag_val;
+}
+
+static u8_t boot_copy_done_get(u32_t bank_offs)
```

Comment:
Do we need to have this function at all? The copy-done is pretty much internal to the bootloader, and shouldn't need to be read here.